src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+
+  close(link.sock);
+  return 0;
+}
+
+static int write_statsd_write(
+    const data_set_t *ds, const value_list_t *vl, user_data_t *ud) {
+  int   idx;
+  _Bool include_ds_name = configuration.always_append_ds || ds->ds_num > 1;
+
+  /*
+   * Build the type name using all available fields in vl:
+   * hostname.plugin[.plugin_instance].type[.type_instance]
+   */
+  char*  value_key = NULL;
+  size_t value_key_len = 2;  // Dots are always present.
+  size_t plugin_instance_len = strlen(vl->plugin_instance);
+  size_t type_instance_len = strlen(vl->type_instance);
+
+  value_key_len += strlen(vl->host);
+  value_key_len += strlen(vl->plugin);
+  value_key_len += strlen(vl->type);
+
+  // Account for name's optional fields.
+  if (plugin_instance_len != 0) {
+    value_key_len += plugin_instance_len + 1;
+  }
+  if (type_instance_len != 0) {
+    value_key_len += type_instance_len + 1;
+  }
+
+  value_key = allocate(value_key_len + 1);
+  if (value_key == NULL) {
+    return -5;
+  }
+
+  // Determine format and build fixed part of the name.
+  if (plugin_instance_len != 0 && type_instance_len != 0) {
+    ssnprintf(value_key, value_key_len + 1, "%s.%s.%s.%s.%s",
+              vl->host, vl->plugin, vl->plugin_instance,
+              vl->type, vl->type_instance);
+  } else if (plugin_instance_len != 0) {
+    ssnprintf(value_key, value_key_len + 1, "%s.%s.%s.%s",
+              vl->host, vl->plugin, vl->plugin_instance, vl->type);
+  } else if (type_instance_len != 0) {
+    ssnprintf(value_key, value_key_len + 1, "%s.%s.%s.%s",
+              vl->host, vl->plugin, vl->type, vl->type_instance);
+  } else {
+    ssnprintf(value_key, value_key_len + 1, "%s.%s.%s",
+              vl->host, vl->plugin, vl->type);
+  }
+
+  // Process all values in the data set.
+  for (idx = 0; idx < ds->ds_num; idx++) {
+    char*  message;
+    size_t message_len;
+    int    result;
+    char*  ds_type = DS_TYPE_TO_STATSD[ds->ds[idx].type];
+    char*  value = NULL;
+
+    if (ds_type == NULL) {
+      if (!configuration.silence_type_warnings) {
+        WARNING("write_statsd plugin: unsupported StatsD type '%s' "
+                "for value with name '%s'.",
+                DS_TYPE_TO_STRING(ds->ds[idx].type), ds->ds[idx].name);
+      }
+      continue;  // To the next value in the data set.
+    }
+
+    value = ds_value_to_string(ds->ds[idx].type, vl->values[idx]);
+    if (value == NULL) {
+      free(value_key);
+      return -6;
+    }
+
+    /*
+     * The full message will have prefix, postfix and separators plus
+     * the type name, the value name, value and type identifier.
+     */
+    if (include_ds_name) {
+      message_len = configuration.event_line_base_len;
+      message_len += strlen(ds->ds[idx].name);
+
+    } else {
+      message_len = configuration.short_event_line_base_len;
+    }
+
+    message_len += value_key_len;
+    message_len += strlen(value);
+    message_len += strlen(ds_type);
+
+    message = allocate(message_len + 1);
```

Comment:
This byte counting needs to go. Use a large enough buffer, e.g. 1 kByte, and go with that. If you're *really* **absolutely** concerned about not wasting a byte of memory, call `snprintf()` twice: once to determine the required buffer size, and a second time with the correctly allocated buffer. Beware, however, that `snprintf()` is *extremely* slow and this will have a large performance impact on this plugin.